### DISCUSSION TITLE
Cannot send BULK vendor response (rp2040)

### DISCUSSION THREAD

**Discussion by GiantNuker** [hathach/tinyusb#706] [Q&A]
I'm trying to send a bulk transfer in response to a control transfer. So far I've managed to send and recieve a control and bulk transfer from the *host*, however I cannot send the transfer  from the *device*.
In the following example, The `Bulk xfer` line should say `OK` and the `Write buffer` line should be back at `64` on the second time i run `refresh`.

https://user-images.githubusercontent.com/43734540/110055867-a1101080-7d12-11eb-8c47-561c5f4c3ce4.mp4

From what I can tell, The transfer never happens, as the write buffer size dosen't return to `64`. Since usb is based on the host polling the device, first I tried running something else on the device, with very similar host code. It worked perfectly fine using [The Pi Pico low level usb example](https://github.com/raspberrypi/pico-examples/blob/master/usb/device/dev_lowlevel/dev_lowlevel.c) when I sent a 64 byte array to it, and it repeated it back. However when I used `tud_vendor_n_write_str` I wasn't able to recieve the bulk response.
The relevant host code:
```c
void refreshConnection() {
    log_message("Refreshing...");

    libusb_device** device_list;
    ssize_t device_count = libusb_get_device_list(NULL, &device_list);
    if (device_count < 0) {
        char resp[100] = "Failed to get device list: ";
        strcpy(resp + 26, libusb_error_name(device_count));
        log_message(resp);
        return;
    }

    for (ssize_t i = 0; i < device_count; i++) {
        struct libusb_device_descriptor desc;
		int r = libusb_get_device_descriptor(device_list[i], &desc);
		if (r < 0) {
			log_message("Failed to get device descriptor");
            continue;
		}

        if (desc.idVendor == 0x1337 && desc.idProduct == 0x400a) {
            log_message("Found keyboard");
            libusb_device_handle* device_handle = NULL;
            libusb_open(device_list[i], &device_handle);
            if (device_handle != 0) {
                log_message("Opened keyboard");
                int r = libusb_claim_interface(device_handle, 2);
                if (r != 0) {
                    char resp[100] = "Failed to claim interface: ";
                    strcpy(resp + 27, libusb_error_name(r));
                    log_message(resp);
                    continue;
                }
                r = libusb_control_transfer(device_handle, 0x41, 0x01, 0, 0, NULL, 0, 1000);
                char resp[100] = "Control xfer: ";
                if (r < 0) {
                    strcpy(resp + 14, libusb_error_name(r));
                } else {
                    strcpy(resp + 14, "OK");
                }
                log_message(resp);
                uint8_t buffer[64] = {0};
                memset(buffer, 0x69, 62);
                int buffer_len = 64;
                r = libusb_bulk_transfer(device_handle, 0x8f, buffer, 64, &buffer_len, 1000);
                char resp2[164] = "Bulk xfer: ";
                if (r < 0) {
                    strcpy(resp2 + 11, libusb_error_name(r));
                } else {
                    strcpy(resp2 + 11, "OK: ");
                }
                log_message(resp2);

                libusb_close(device_handle);
                //todo
            } else {
                log_message("Unable to open keyboard");
            }
        } else {
            // char msg[100] = "Found non-keyboard: ";
            // sprintf(msg + 20, "VID: %04x PID: %04x", desc.idVendor, desc.idProduct);
            // log_message(msg);
        }
    }

    libusb_free_device_list(device_list, 1);
}
```

The relevant device code:
```c
bool tud_vendor_control_request_cb(uint8_t rhport, tusb_control_request_t const * request) {
    switch (request->bRequest) {
        case 0x01:
            gpio_put(PIN_LED3, true);
            printf("Aight\n");
            printf("Write buffer size: %d\n", tud_vendor_n_write_available(USB_VENDOR_DEBUG_N));
            tud_vendor_n_write_str(USB_VENDOR_DEBUG_N, "respondssssseee!");
            printf("Write buffer size: %d\n", tud_vendor_n_write_available(USB_VENDOR_DEBUG_N));
            return tud_control_status(rhport, request);
        default:
            return false;
    }
}
```
Anyone have a clue?
I looked at the webusb example and picoprobe, which both appear to use it exactly the same way without any issues.

* Yes I have made sure I'm using the correct interfaces/endpoints (I checked all of them just to be sure)
* Yes tud_task() is being called

**Comment by hathach** [comment]
try to build tinyusb with CFG_TUSB_DEBUG=2 and have your printf() output to uart, then post the tinyusb debug log here. I could have a hint where thing failed. Also make sure you build with tinyusb latest code, there are lots of fixes lately for rp2040 port.

**Reply by GiantNuker** [reply]
Updating submodules seems to have fixed it
```
USBD Setup Received 41 01 00 00 00 00 00 00 
Aight
Write buffer size: 64
  Queue EP 8F with 16 bytes ... OK
Write buffer size: 64
  Queue EP 80 with zlp Status
USBD Xfer Complete on EP 80 with 0 bytes
USBD Xfer Complete on EP 8F with 16 bytes
  VENDOR xfer callback
```

**Reply by hathach** [reply]
FYI #609  #645 or #646

**Reply by GiantNuker** [reply]
@hathach just commnted the flag and it stopped working; `CFG_TUSB_DEBUG=2` fixes the issue though.

**Reply by hathach** [reply]
You seem to have incorrect conifg descriptor the log show endpoint 0x8f which is a bit odd.

**Reply by GiantNuker** [reply]
0x8f is intentional, that way I can add more interfaces without having to rearrange the debug one so its last.

**Reply by hathach** [reply]
If you are on Linux, try to update udev rules accordinly https://github.com/hathach/tinyusb/pull/710 . Other than that i don't see any other issues in your log. It is probably due to your own firmware implementation, try to throw a bit of printf() and step through the program to troubleshoot. Keep updating this thread if you figure out something, maybe other with similar set up could help.
